driver/src/main/com/mongodb/GridFSBucketImpl.java

```diff
@@ -0,0 +1,316 @@
+        }
+    }
+
+    @Override
+    public void rename(final ObjectId id, final String newFilename) {
+        UpdateResult updateResult = filesCollection.updateOne(new BsonDocument("_id", new BsonObjectId(id)),
+                new BsonDocument("$set", new BsonDocument("filename", new BsonString(newFilename))));
+
+        if (updateResult.wasAcknowledged() && updateResult.getMatchedCount() == 0) {
+            throw new MongoGridFSException(format("No file found with the ObjectId: %s", id));
+        }
+    }
+
+    private MongoCollection<Document> getFilesCollection() {
+        return getFilesCollection(Document.class);
+    }
+
+    private <TResult> MongoCollection<TResult> getFilesCollection(final Class<TResult> clazz) {
+        return database.getCollection(bucketName + ".files", clazz)
+                .withCodecRegistry(codecRegistry)
+                .withReadPreference(readPreference)
+                .withWriteConcern(writeConcern);
+    }
+
+    private <TResult> MongoCollection<TResult> getChunksCollection(final Class<TResult> clazz) {
+        return database.getCollection(bucketName + ".chunks", clazz)
+                .withCodecRegistry(codecRegistry)
+                .withReadPreference(readPreference)
+                .withWriteConcern(writeConcern);
+    }
+
+    private void checkCreateIndex() {
+        if (!checkedIndexes) {
+            if (filesCollection.withReadPreference(primary()).find()
+                    .projection(new BsonDocument("_id", new BsonInt32(1))).first() == null) {
+                filesCollection.createIndex(Indexes.ascending("filename", "uploadDate"));
+                chunksCollection.createIndex(Indexes.ascending("files_id", "n"), new IndexOptions().unique(true));
+            }
+            checkedIndexes = true;
+        }
+    }
+
+    private BsonDocument getFileByName(final String filename, final GridFSDownloadByNameOptions options) {
+        int revision = options.getRevision();
+        int skip;
+        BsonInt32 sort;
+        if (revision >= 0) {
+            skip = revision;
+            sort = new BsonInt32(1);
+        } else {
+            skip = (-revision) - 1;
+            sort = new BsonInt32(-1);
+        }
+
+        BsonDocument fileInfo = filesCollection.find(new BsonDocument("filename", new BsonString(filename)))
+                .skip(skip).sort(new BsonDocument("uploadDate", sort)).first();
+        if (fileInfo == null) {
+            throw new MongoGridFSException(format("No file found with the filename: %s and revision: %s", filename, revision));
+        }
+        return fileInfo;
+    }
+
+    private void downloadToStream(final GridFSDownloadStream stream, final OutputStream destination) {
+        byte[] buffer = new byte[stream.getFileInformation().getInt32("chunkSize").getValue()];
+        int len;
+        try {
+            while ((len = stream.read(buffer)) != -1) {
```

Comment:
See above note about the behavior of InputStream.read  It's likely ok here, since we know the implementation of GridFSDownloadStream, but it is fragile.